### PR TITLE
Unified runtime for Play Lab

### PR DESCRIPTION
**Pull request by joshlory**
Follow up to PR https://github.com/code-dot-org/code-dot-org/pull/10152. This migrates Play Lab to a unified runtime.
- [x] Fix multiple events with the same name.
- [x] Give success/failure conditions access to interpreted globals.
- [x] Update the `studio_ask` callbacks to be aware of the interpreter.

### CODE REVIEW

File: apps/src/codegen.js [hunk 1]
```diff
@@ -52,23 +52,29 @@ exports.evalWith = function (code, options, legacy) {
...
-exports.evalWithEvents = function (apis, events, generator) {
+exports.evalWithEvents = function (apis, events, evalCode = '') {
   let interpreter, currentCallback;
-  let code = '';
   const hooks = {};
   Object.keys(events).forEach(event => {
-    hooks[event] = () => {
-      currentCallback(event);
+    hooks[event] = (...args) => {
+      const eventArgs = {name: event, args: args};
```
Comment (Bjvanminnen) [inline apps/src/codegen.js]: Could just be:
`const eventArgs = {name: event, args};`

File: apps/src/codegen.js [hunk 2]
```diff
@@ -52,23 +52,29 @@ exports.evalWith = function (code, options, legacy) {
...
+      currentCallback(exports.marshalNativeToInterpreter(interpreter, eventArgs, null, 5));
       interpreter.run();
     };
-    code += `function ${event}(){${generator(events[event])}};`;
+    const {code, args} = events[event];
+    evalCode += `this['${event}']=function(${args ? args.join() : ''}){${code}};`;
```
Comment (Bjvanminnen) [inline apps/src/codegen.js]: I think args will always be truthy here (and an empty array in the case where no args are provided)

File: apps/src/studio/studio.js [hunk 3]
```diff
@@ -234,6 +234,9 @@ function loadLevel() {
   }
   blocks.registerCustomGameLogic(Studio.customLogic);
 
+  Studio.legacyRuntime = Studio.customLogic;
+
+
```
Comment (Bjvanminnen) [inline apps/src/studio/studio.js:239]: extra line?

File: apps/src/studio/studio.js [hunk 4]
```diff
@@ -234,6 +234,9 @@ function loadLevel() {
   }
   blocks.registerCustomGameLogic(Studio.customLogic);
 
+  Studio.legacyRuntime = Studio.customLogic;
```
Comment (Bjvanminnen) [inline apps/src/studio/studio.js]: worth a comment explaining what it means to be a legacy runtime/why we still need to support it/if we have a plan to get rid of it entirely at some point?

File: apps/src/studio/studio.js [hunk 5]
```diff
@@ -2950,9 +2963,24 @@ Studio.execute = function () {
...
     // (must be after reset(), which resets the Studio.Globals namespace)
-    defineProcedures('procedures_defreturn');
-    defineProcedures('procedures_defnoreturn');
-    defineProcedures('functional_definition');
+    if (Studio.legacyRuntime) {
+      defineProcedures('procedures_defreturn');
+      defineProcedures('procedures_defnoreturn');
+      defineProcedures('functional_definition');
+    } else {
+      var generator = Blockly.Generator.blockSpaceToCode.bind(Blockly.Generator, 'JavaScript');
```
Comment (Bjvanminnen) [inline apps/src/studio/studio.js]: could be const

File: apps/src/studio/studio.js [hunk 6]
```diff
@@ -2950,9 +2963,24 @@ Studio.execute = function () {
...
+      const code = [
+        'procedures_defreturn',
+        'procedures_defnoreturn',
+        'functional_definition'
+      ].map(p => generator(p)).join(';');
```
Comment (Hamms) [inline apps/src/studio/studio.js]: could this just be `.map(generator)`?

**Comment by islemaster** [comment]
LGTM.

**Comment by Hamms** [comment]
LGTM!

File: apps/src/studio/studio.js [hunk 7]
```diff
@@ -4879,7 +4924,28 @@ Studio.askForInput = function (question, callback) {
 
     let handlerName = `askCallback${Studio.askCallbackIndex}`;
     Studio.askCallbackIndex++;
-    registerEventHandler(Studio.eventHandlers, handlerName, callback.bind(null, value || ''));
+    // Shift a CallExpression node on the stack that already has its func_,
+    // arguments, and other state populated:
```
Comment (joshlory) [inline apps/src/studio/studio.js:4929]: This is the same idea as [`codegen.createNativeInterpreterCallback`](https://github.com/code-dot-org/code-dot-org/blob/272f776281e472610d82495376a9b41337f050b7/apps/src/codegen.js#L397-L440) but it only executes until we get back to the starting depth then re-pauses the interpreter.

The plan is for this hack to go away once each callback runs in a separate JSInterpreter instance.

**Comment by joshlory** [comment]
@Hamms @islemaster @Bjvanminnen @cpirich I've updated this PR to fix tests and a few other issues that surfaced.  Can you take another look when you get a chance?

Thanks!

**Comment by islemaster** [comment]
LGTM!

**Comment by joshlory** [comment]
I've checked the following Play Lab stages:
- /s/course1 stage 16
- /s/course2 stage 17
- /s/course3 stages 16 & 17
- /s/course4 stages 7 & 11

I found one issue that affects 4.7.7, 4.11.4 & 4.11.5.  Multiple event handlers of the same type are being run in sequence instead of in parallel.  This side-by-side gif illustrates the issue:

studio

I still need to check /s/course4 stage 15.  4.11.6 has an unrelated (existing) issue that is being fixed on Levelbuilder.

**Comment by ashercodeorg** [comment]
In the right GIF, the code highlight is only highlighting the last `repeat forever`. Is it possible to have all them highlight appropriately?

**Comment by joshlory** [comment]
In the future, yes 😄.  The right gif is the existing behavior.  I think it's a limitation of Blockly that it can only highlight one block at a time (setting a new highlight removes the existing highlight).

**Comment by joshlory** [comment]
I believe I now have the parallel queues running correctly.  I'm going to take one final pass through the Play Lab levels in Courses 1-4.  Barring any issues I plan to ship this for Tuesdays DTP.

Thank you everyone for the review!

**Comment by Hamms** [comment]
LGTM!

**Pull request merged**